File: 67-AddBinary/67-AddBinary.java
// Last updated: 8/16/2026, 9:53:00 PM
class Solution {
    public String addBinary(String a, String b) {
        char[] aChars = a.toCharArray();
        char[] bChars = b.toCharArray();
        StringBuilder res = new StringBuilder();

        int i = a.length() - 1, j = b.length() - 1;
        int carry = 0;
        while( i >= 0 || j >= 0) {
            int sum = carry;
            if(i >= 0) sum += aChars[i --] - '0';
            if(j >= 0) sum += bChars[j --] - '0';
            
            // if(sum < 2) {
            //     res.append(sum);
            //     carry = 0;
            // } else {
            //     res.append(sum - 2);
            //     carry = 1;
            // }

            // OR use the binary feature, relation of 2
            res.append( sum % 2 );
            carry = sum / 2;
        }
        if(carry > 0) res.append(carry);

        return res.reverse().toString();

    }
}

File: 67-AddBinary/67-AddBinary.py
# Last updated: 9/1/2026, 3:43:09 PM
1class Solution:
2    def addBinary(self, a: str, b: str) -> str:
3        # start from the end, then reverse
4        i, j = len(a)-1, len(b)-1
5        carry = 0
6        res = []   # list of str of added res, in reverse order
7
8        while i >= 0 or j >= 0 or carry > 0:
9            total = carry   # total will be reset to carry every round
10
11            if i >= 0:
12                total += int(a[i])
13                i -= 1
14            if j >= 0:
15                total += int(b[j])
16                j -= 1
17        
18            total, carry = total % 2, total // 2
19            res.append(str(total))
20        
21        return ''.join(reversed(res))
22